File: backend/src/main/java/com/github/aadvorak/artilleryonline/battle/calculator/missile/CorrectingAccelerationCalculator.java
package com.github.aadvorak.artilleryonline.battle.calculator.missile;

import com.github.aadvorak.artilleryonline.battle.calculations.BattleCalculations;
import com.github.aadvorak.artilleryonline.battle.calculations.MissileCalculations;
import com.github.aadvorak.artilleryonline.battle.calculator.TargetCalculator;
import com.github.aadvorak.artilleryonline.battle.specs.MissileSpecs;
import com.github.aadvorak.artilleryonline.battle.utils.GeometryUtils;

import java.util.stream.Collectors;

public class CorrectingAccelerationCalculator {

    private static final double PRECISION_THRESHOLD = 6.0;

    private static final double CLOSE_EDGE_DISTANCE = 5.0;

    private static final double CLOSE_EDGE_ANGLE = - Math.PI / 4;

    private static final double FAR_EDGE_DISTANCE = 10.0;

    private static final double FAR_EDGE_ANGLE = Math.PI / 5;

    private static final double K = (FAR_EDGE_ANGLE - CLOSE_EDGE_ANGLE) / (FAR_EDGE_DISTANCE - CLOSE_EDGE_DISTANCE);

    private static final double B = CLOSE_EDGE_ANGLE - CLOSE_EDGE_DISTANCE * K;

    public static void calculate(MissileCalculations calculations, BattleCalculations battle) {
        var missileState = calculations.getModel().getState();
        var missileSpecs = calculations.getModel().getSpecs();
        var velocityMagnitude = missileState.getVelocity().getMovingVelocity().magnitude();
        var correctingVelocity = velocityMagnitude - missileSpecs.getMinCorrectingVelocity();
        var missilePosition = missileState.getPosition();
        if (correctingVelocity <= 0) {
            calculations.setCorrectingAcceleration(
                    getAcceleration(missilePosition.getAngle(), Math.PI / 2, missileSpecs, velocityMagnitude)
            );
            return;
        }
        var targets = TargetCalculator.calculatePositions(calculations.getModel().getVehicleId(), battle);
        if (targets.isEmpty()) {
            return;
        }
        var targetDataSet = targets.stream()
                .map(position -> new TargetData(missilePosition.getCenter().angleTo(position),
                        missilePosition.getCenter().distanceTo(position),
                        position.getX() - missilePosition.getX()))
                .collect(Collectors.toSet());
        var iterator = targetDataSet.iterator();
        var closestTarget = iterator.next();
        while (iterator.hasNext()) {
            var targetData = iterator.next();
            if (Math.abs(targetData.xDistance) < Math.abs(closestTarget.xDistance)) {
                closestTarget = targetData;
            }
        }
        var absXDistance = Math.abs(closestTarget.xDistance);
        if (absXDistance > FAR_EDGE_DISTANCE) {
            var tagetAngle = closestTarget.xDistance > 0 ? FAR_EDGE_ANGLE : Math.PI - FAR_EDGE_ANGLE;
            calculations.setCorrectingAcceleration(
                    getAcceleration(missilePosition.getAngle(), tagetAngle, missileSpecs, velocityMagnitude)
            );
            return;
        }
        if (absXDistance > CLOSE_EDGE_DISTANCE) {
            var angle = K * absXDistance + B;
            var tagetAngle = closestTarget.xDistance > 0 ? angle : Math.PI - angle;
            calculations.setCorrectingAcceleration(
                    getAcceleration(missilePosition.getAngle(), tagetAngle, missileSpecs, velocityMagnitude)
            );
            return;
        }
        var angleDiff = GeometryUtils.calculateAngleDiff(missilePosition.getAngle(), closestTarget.angle);
        if (Math.abs(angleDiff) < missileSpecs.getAnglePrecision()
                * getAnglePrecisionCoefficient(closestTarget.distance)) {
            return;
        }
        calculations.setCorrectingAcceleration(Math.signum(angleDiff) * correctingVelocity
                * missileSpecs.getCorrectingAccelerationCoefficient());
    }

    private static double getAcceleration(double missileAngle, double targetAngle,
                                          MissileSpecs missileSpecs, double velocityMagnitude) {
        return Math.signum(GeometryUtils.calculateAngleDiff(missileAngle, targetAngle)) * velocityMagnitude
                * missileSpecs.getCorrectingAccelerationCoefficient() / missileSpecs.getMinCorrectingVelocity();
    }

    private static double getAnglePrecisionCoefficient(double distance) {
        if (distance < PRECISION_THRESHOLD) {
            return distance /  PRECISION_THRESHOLD;
        }
        return 1.0;
    }

    private record TargetData(double angle, double distance, double xDistance) {
    }
}

File: frontend/playground/battle/calculator/missile/correcting-acceleration-calculator.ts
import type {MissileModel} from "~/playground/data/model";
import {VectorUtils} from "~/playground/utils/vector-utils";
import {BattleUtils} from "~/playground/utils/battle-utils";
import {TargetCalculator} from "~/playground/battle/calculator/target-calculator";
import type {MissileSpecs} from "~/playground/data/specs";
import type {BattleCalculations} from "~/playground/data/calculations";

const PRECISION_THRESHOLD = 6.0
const CLOSE_EDGE_DISTANCE = 5.0
const CLOSE_EDGE_ANGLE = -Math.PI / 4
const FAR_EDGE_DISTANCE = 10.0
const FAR_EDGE_ANGLE = Math.PI / 5

const K = (FAR_EDGE_ANGLE - CLOSE_EDGE_ANGLE) / (FAR_EDGE_DISTANCE - CLOSE_EDGE_DISTANCE)
const B = CLOSE_EDGE_ANGLE - CLOSE_EDGE_DISTANCE * K

class TargetData {
  constructor(
      public angle: number,
      public distance: number,
      public xDistance: number
  ) {}
}

export const CorrectingAccelerationCalculator = {

  calculate(missileModel: MissileModel, battle: BattleCalculations): number {
    const state = missileModel.state
    const specs = missileModel.specs
    const velocityMagnitude = VectorUtils.getMagnitude(state.velocity)
    const correctingVelocity = velocityMagnitude - specs.minCorrectingVelocity

    const missilePosition = state.position

    if (correctingVelocity <= 0) {
      return this.getAcceleration(
          missilePosition.angle,
          Math.PI / 2,
          specs,
          velocityMagnitude
      )
    }

    const targets = TargetCalculator.calculatePositions(missileModel.vehicleId, battle)
    if (!targets.length) {
      return 0.0
    }

    const targetDataSet = new Set<TargetData>()
    for (const position of targets) {
      const angle = VectorUtils.angleFromTo(missilePosition, position)
      const distance = BattleUtils.distance(missilePosition, position)
      const xDistance = position.x - missilePosition.x
      targetDataSet.add(new TargetData(angle, distance, xDistance))
    }

    let closestTarget = null as TargetData | null
    for (const td of targetDataSet) {
      if (!closestTarget) {
        closestTarget = td
      } else if (Math.abs(td.xDistance) < Math.abs(closestTarget.xDistance)) {
        closestTarget = td
      }
    }

    if (!closestTarget) {
      return 0.0
    }

    const absXDistance = Math.abs(closestTarget.xDistance)

    if (absXDistance > FAR_EDGE_DISTANCE) {
      const tagetAngle = closestTarget.xDistance > 0 ? FAR_EDGE_ANGLE : Math.PI - FAR_EDGE_ANGLE
      return CorrectingAccelerationCalculator.getAcceleration(
          missilePosition.angle,
          tagetAngle,
          specs,
          velocityMagnitude
      )
    }

    if (absXDistance > CLOSE_EDGE_DISTANCE) {
      const angle = K * absXDistance + B
      const tagetAngle = closestTarget.xDistance > 0 ? angle : Math.PI - angle
      return CorrectingAccelerationCalculator.getAcceleration(
          missilePosition.angle,
          tagetAngle,
          specs,
          velocityMagnitude
      )
    }

    const angleDiff = BattleUtils.calculateAngleDiff(missilePosition.angle, closestTarget.angle)
    if (Math.abs(angleDiff) < specs.anglePrecision
        * CorrectingAccelerationCalculator.getAnglePrecisionCoefficient(closestTarget.distance)) {
      return 0.0
    }

    return Math.sign(angleDiff) * correctingVelocity * specs.correctingAccelerationCoefficient
  },

  getAcceleration(
      missileAngle: number,
      targetAngle: number,
      specs: MissileSpecs,
      velocityMagnitude: number
  ): number {
    return Math.sign(BattleUtils.calculateAngleDiff(missileAngle, targetAngle)) * velocityMagnitude
        * specs.correctingAccelerationCoefficient / specs.minCorrectingVelocity
  },

  getAnglePrecisionCoefficient(distance: number): number {
    if (distance < PRECISION_THRESHOLD) {
      return distance / PRECISION_THRESHOLD
    }
    return 1.0
  }
}
